src/reverse_proxy_handler.cc

```diff
@@ -0,0 +1,206 @@
+#include "reverse_proxy_handler.h"
+#include "config_parser.h"
+#include <iostream>
+#include <boost/asio.hpp>
+
+
+RequestHandler::Status ReverseProxyHandler::Init(const std::string& uri_prefix, const NginxConfig& config) {
+    prefix_ = uri_prefix;
+    host_ = "";
+    urlpath_ = "";
+    port_ = "";
+
+    std::string fullHost = "";
+    for (auto statement : config.statements_){
+      if (statement->tokens_.size() > 1 && statement->tokens_[0] == "port"){
+	port_ = statement->tokens_[1];
+      }
+      else if (statement->tokens_.size() > 1 && statement->tokens_[0] == "host"){
+        fullHost = statement->tokens_[1];
+      }
+    }
+    
+    if (fullHost == "")
+      return RequestHandler::Status::INVALID_CONFIG;
+
+    //Just a host is provided in the config
+    if (fullHost.find("/") == std::string::npos){
+      host_ = fullHost;
+      urlpath_ = "/";
+    }
+    else {
+      std::size_t hostEnd_pos = fullHost.find("/");
+      host_ = fullHost.substr(0, hostEnd_pos);
+      urlpath_ = fullHost.substr(hostEnd_pos, fullHost.size() - host_.size());
+    }
+    if (urlpath_.back() != '/')
+      urlpath_ += "/";
+
+    if (port_ == ""){
+      port_ = "80";
+    }
+
+    return RequestHandler::Status::OK;
+}
+
+RequestHandler::Status ReverseProxyHandler::HandleRequest(const Request& request, Response* response) {
+     
+    const int MAX_REDIRECTS = 21;
+    //Transform request to be forwarded to provided host
+    //TODO: raw_request no longer matches other request member - fix that?
+    auto transformedRequest = TransformRequest(request);
+    
+    //Use a host parameter for ForwardRequest instead of updating host_ each time
+    std::string nextHost = host_;
+    //Now loop for maximum redirects
+
+    for (int i = 0; i < MAX_REDIRECTS; i++){
+      std::cout << "Beginning to forward request in ReverseProxyHandler to " << host_ << std::endl;
+      
+      Response::ResponseCode forwardRC = ForwardRequest(transformedRequest, response, nextHost);
```

Comment:
I find it a little weird that ForwardRequest returns a Response Code that does not correspond with the actual response code, which you check later. Would it be more natural to return a bool indicating whether or not the request was forwarded properly (since ForwardRequest returns either OK or Internal Server Error). Another option would be to return a RequestHandler Status instead (either OK or Proxy Error). Or you could return the actual response code, and then check if Internal Server Error was returned.